hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -250,49 +333,33 @@ class ServiceBackend(
     }
   }
 
-  def execute(username: String, sessionID: String, billingProject: String, bucket: String, code: String, token: String): String = {
-    ExecutionTimer.logTime("ServiceBackend.execute") { timer =>
-      userContext(username, timer) { ctx =>
-        log.info(s"executing: ${token}")
-        ctx.backendContext = new ServiceBackendContext(username, sessionID, billingProject, bucket)
-
-        execute(ctx, IRParser.parse_value_ir(ctx, code)) match {
-          case Some((v, t)) =>
-            JsonMethods.compact(
-              JObject(List("value" -> JSONAnnotationImpex.exportAnnotation(v, t.virtualType),
-                "type" -> JString(t.virtualType.toString))))
-          case None =>
-            JsonMethods.compact(
-              JObject(List("value" -> null, "type" -> JString(TVoid.toString))))
-        }
-      }
+  def execute(
+    tmpdir: String,
+    sessionId: String,
+    billingProject: String,
+    remoteTmpDir: String,
+    code: String,
+    token: String
+  ): String = serviceBackendExecuteContext(
+    "ServiceBackend.execute",
+    tmpdir,
+    sessionId,
+    billingProject,
+    remoteTmpDir
+  ) { ctx =>
+    log.info(s"executing: ${token}")
+
+    execute(ctx, IRParser.parse_value_ir(ctx, code)) match {
+      case Some((v, t)) =>
+        JsonMethods.compact(
+          JObject(List("value" -> JSONAnnotationImpex.exportAnnotation(v, t.virtualType),
```

Comment:
Fine for now, but we should switch to the encoded protocol John added pretty much immediately after this goes in.

Reply:
Agreed, and it shouldn't take long to switch I think

Reply:
I have marked this in my todo.